src/ray/core_worker/core_worker.cc

```diff
@@ -1895,12 +1895,19 @@ void CoreWorker::HandlePushTask(const rpc::PushTaskRequest &request,
     return;
   }
 
+  // Enqueue the task
   task_queue_length_ += 1;
+  direct_task_receiver_->EnqueuePushedTask(request, reply, send_reply_callback);
+
+  // Extract the caller's WorkerID
+  const TaskSpecification task_spec(request.task_spec());
+  WorkerID caller_worker_id = task_spec.CallerWorkerId();
+
   task_execution_service_.post([=] {
     // We have posted an exit task onto the main event loop,
     // so shouldn't bother executing any further work.
     if (exiting_) return;
-    direct_task_receiver_->HandlePushTask(request, reply, send_reply_callback);
+    direct_task_receiver_->RunTasksFromQueue(caller_worker_id);
```

Comment:
Hmm this seems to break the whole thing? Doesn't this mean we won't use `HandlePushTask` method anymore? But it doesn't look like `RunTasksFromQueue` can completely replace that method.

Reply:
Yes! We want to separate the former `HandlePushTask` function in two parts: one that enqueues tasks (`CoreWorkerDirectTaskReceiver::EnqueuePushedTask`) to the appropriate worker's queue, and one that executes them (`CoreWorkerDirectTaskReceiver::RunTasksFromQueue`). The former is run by a worker as soon as it receives a new task (in CoreWorker::HandlePushTask), while the other is posted to the worker's task_execution_service.